# 1. 다시 보는 초난감 DAO

- UserDao는 아직 예외상황에 대한 처리가 미흡하다.

## 1.1 예외처리 기능을 갖춘 DAO

- DB 커넥션이라는 제한적인 리소스를 공유해 사용하는 서버에서 동작하는 JDBC 코드는 예외처리에 대한 지켜야 할 원칙이 있다.
    - 어떤 예외가 발생하더라도 반드시 리소스를 반환하도록 만들어야 한다.

### JDBC 수정 기능의 예외처리 코드

- `deleteAll()`은 Connection과 PreparedStatement라는 두 개의 공유 리소스를 가져와 사용한다.
- 예외가 발생하면 `close()`를 실행하지 못하고 메서드를 빠져나간다. 즉 리소스 반환이 제대로 이뤄지지 않을 수 있다.
- 일반적으로 서버에서는 제한된 개수의 DB 커넥션을 만들어서 재사용 가능한 풀로 관리한다.
- DB 풀은 매번 명시적으로 close() 해서 돌려줘야지만 재사용 가능하다.
- 에러가 나서 반환하지 못하는 Connection이 쌓이면 리소스가 모자란다는 심각한 오류를 내며 서버가 중단될 수 있다.
- 그래서 이런 JDBC 코드에서는 `try-catch-finally` 구문이 권장된다.

```java
public class UserDao {
    // ...
    public void deleteAll() throws SQLException {
        Connection c = null;
        PreparedStatement ps = null;

        try {
            c = dataSource.getConnection();
            ps = c.prepareStatement("DELETE FROM users");
            ps.execute();
        } catch (Exception e) {
            throw e;
        } finally {
            if (ps != null) {
                try {
                    ps.close();
                } catch (SQLException e) {
                    // pass
                }
            }

            if (c != null) {
                try {
                    c.close();
                } catch (SQLException e) {
                    // pass
                }
            }
        }
    }
    // ...
}
```

---
[목록](./index.md)
